### ISSUE TITLE
Discussion for tasker functions

### ISSUE THREAD

**Issue by victorbalan** [coduno/api#51]
The following will describe how coding speed is defined. Coding speed is referring to the number of tries a coder had until he submitted the final green result and also the time in which he did it.

The codingSpeed has a value between 0 and 1. From now on i will use `matlab` notations.
So, `codingSpeed:=[0..1]`. It will be computed from a number of tests.
 * The first test will take in account the number of submissions until the final
result.
 * The second test will take in account the number of inserted/deleted lines
until the final result.
 * The third test will take in account the time until the final result.

### Coding speed formula

**Legend**
       * cs=codingSpeed
       * nos=number of sumbissions test
       * idl=inserted/deleted lines test
       * timet=time test

  `cs = nosWeight * nos + idlWeight * idl + timetWeight * timet`

where `nosWeight + idlWeight + timetWeight=1`

### Function definitions

* nos will be defined by a function `f(s,t)` where s represents number of submissions
and t represents total number of tests. If `f(s,t)<0` then the coder overpassed
the max number of submissions.

            `f(s,t):=1-s/t*10`

* timet will be defined by a function `f(t,max)` where `t` represents the time in
wich the coder finished the task and `max` represents the max allowed time for that task.

            `f(t,max):=t/max`

* idl -> to be defined

**Comment by flowlo** [comment]
You define `f` twice, and this is not MATLAB. So, to put this in *real* MATLAB syntax, you mean:

```matlab
function x = nOfSubmissionsTest(s, t)
	% s is the number of submissions the user made
	% t is the total number of tests
	x = 1 - s / t * 10
end

function x = timeTest(t, max)
	% t is the time at which the user submitted the final submissions
	% max is the maximum allowed time for the task
	x = t / max
end

function x = idl(x)
	x = x % to be defined
end

function x = cs(s, t1, t2, max, nosWeight, timetWeight, idlWeight)
	if nosWeight + timetWeight + idlWeight ~= 1
		% weights do not sum up, return something insane ...
		x = -Inf
		return
	end
	x = nOfSubmissionsTest(s, t1) * nosWeight + timeTest(t2, max) * timetWeight
end
```

Where `cs` gives the "Coding Speed". Is that correct?

**Comment by pbochis** [comment]
For the nrOfSubmissionsTest are you missing some paranthesis? Shouldn/'t it be x = 1 - s / (t * 10) ? Because in the current way, more often than not the result will be f(s,t) < 0

**Comment by victorbalan** [comment]
I updated my first comment

**Comment by flowlo** [comment]
@pbochis Yeah, probably I missed that when transcribing @victorbalan's dialect of what he says is MATLAB to valid MATLAB syntax, sorry. Was not on purpose. I think the convention of expressing ourself in MATLAB makes great sense, because we can all follow. So what you propose is:

```matlab
function x = nOfSubmissionsTest(s, t)
	% s is the number of submissions the user made
	% t is the total number of tests
	if s = 1
		x = 1
		return
	end
	x = 1 - s / (t * 10)
end
```

I am not sure about the whole `nOfSubmissions` thing, because (at least to my understanding) it does not take into account how these submissions turned out (e.g. whether the "tests" said that they are valid). If someone submits multiple valid solutions all the time, we give lower score. maybe the coders just improved the performance or something the like. Also, why is this magic 10 in there? I view it as some threshold for how many tries are acceptable per "test". Is that correct? @victorbalan  what was your intention? Probably we should parameterize this.

For `timeTest` we currently have the inverse of what makes sense.

@victorbalan How would you make `insertedDeletedLinesTest` compare between submissions for different tasks? Also the common scenario is some inserted lines for the first submission, then probably changes, which result in an equal amount of insertions and deletions. So we basically get the number of initial lines, diluted by some factor of change. We'll have to test that.

**Comment by flowlo** [comment]
By the way, why did you decide to introduce a new skill, instead of [one that we already agreed upon]8http://godoc.org/github.com/coduno/api/model#Skills)?

**Comment by victorbalan** [comment]
TODO(flowlo, victorbalan): Add further categories of assessment.

**Comment by flowlo** [comment]
Fair enough. It's just that we're both in the `TODO` ...

**Comment by pbochis** [comment]
I do not agree with how the code speed formula is calculated. 
For one solution, this
```matlab
    x = nOfSubmissionsTest(s, t1) * nosWeight + timeTest(t2, tmax) * timetWeight
```

Should be replaced with this
```matlab
    x = nOfSubmissionsTest(s, t1) * nosWeight - timeTest(t2, tmax) * timetWeight
```
because the lower the timeTest is, the better the coder did(he finished the task faster). If it's high, he should be penalised more. 

A second solution, is leave the main formula the same and modift timeTest().

**Comment by flowlo** [comment]
@pbochis That's what I meant when I said

> For `timeTest` we currently have the inverse of what makes sense.

I think those testing functions should always return higher values for better skills.

**Comment by victorbalan** [comment]
It`s my bad. I had it in atom like this

```matlab
function x = timeTest(d, tmax)
% x is between [0..1]
% d is the time it took the coder to complete the task
% tmax is the maximum allowed time for the task
	x = 1 - t / (tmax * 2)
end
```

**Comment by flowlo** [comment]
Looks better.
